vm/parser.ts

```diff
@@ -693,7 +694,50 @@ module J2ME {
     return "$" + StringUtilities.variableLengthEncodeInt32(hash);
   }
 
+  /**
+   * Encodes variable length utf8 alpha strings of the form [a-z]* to
+   * 32 bit numbers. Below are some sample encodings:
+   *
+   *  "" => 0
+   *  "a" => 1
+   *  "b" => 2 ...
+   *  "z" => 26
+   *  "aa" => 27
+   *  "ab" => 28 ...
+   *  "zz" => 703
+   *  "aaa" => 704
+   *  "azz" => 1378
+   *  "zzz" => 18278
+   *
+   *  The goal of this encoding is to map short strings to low numeric values
+   *  that we can then use to index into tables.
+   */
+  export function lowerCaseAlphaToInt32(utf8String: Uint8Array): number {
+    // We can't encode strings larger than 6 characters because we don't
+    // have enough bits. Technically the limit is somewhere between 6 and 7
+    // but we don't bother to check that here.
+    if (utf8String.length > 6) {
+      // It's okay to return |-1| as a fail value since we'll never use the
+      // highest order bit for encoding.
+      return -1;
+    }
+    var s = 0;
+    for (var i = 0; i < utf8String.length; i++) {
+      var v = utf8String[i] - UTF8Chars.a;
+      if (v < 0 || v >= 26) { // Only 'a' ... 'z' is allowed.
+        return -1;
+      }
+      s *= 26;
+      s += (1 + v); // 'a' is mapped to 1.
+    }
+    return s;
+  }
+
   export function mangleClassName(utf8Name: Uint8Array) {
+    var hash = lowerCaseAlphaToInt32(utf8Name);
+    if (hash > 0 && hash < 2048) {
+      return "$" + fromUTF8(utf8Name);
```

Comment:
I don't think so. Concat functions as far as I understand only help with
reducing the allocation on intermediate strings.
On Apr 21, 2015 4:17 PM, "Marco" <notifications@github.com> wrote:

> In vm/parser.ts
> <https://github.com/mozilla/j2me.js/pull/1425#discussion_r28832012>:
>
> >    export function mangleClassName(utf8Name: Uint8Array) {
> > +    var hash = lowerCaseAlphaToInt32(utf8Name);
> > +    if (hash > 0 && hash < 2048) {
> > +      return "$" + fromUTF8(utf8Name);
>
> Is it worth to create a *concat2* method for this?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mozilla/j2me.js/pull/1425/files#r28832012>.
>